handler/incoming_instance_handler.go

```diff
@@ -108,25 +110,56 @@ func (hdlr *InstanceEventHandler) insertDimensions(ctx context.Context, instance
 
 	cache := make(map[string]string)
 
+	var wg sync.WaitGroup
+	problem := make(chan error, len(dimensions))
+	count := 0
+
 	for _, dimension := range dimensions {
+		wg.Add(1)
+		count++
+		go func(d *model.Dimension) {
+			defer wg.Done()
+
+			dbDimension, err := hdlr.Store.InsertDimension(ctx, cache, instance.DbModel().InstanceID, d.DbModel())
+			if err != nil {
+				err = errors.Wrap(err, "error while attempting to insert dimension")
+				log.Event(ctx, err.Error(), log.Error(err), log.Data{"instance_id": instance.DbModel().InstanceID, "dimension_id": dbDimension.DimensionID})
+				problem <- err
+				return
+			}
 
-		dbDimension, err := hdlr.Store.InsertDimension(ctx, cache, instance.DbModel().InstanceID, dimension.DbModel())
-		if err != nil {
-			return errors.Wrap(err, "error while attempting to insert dimension")
-		}
+			if err = hdlr.DatasetAPICli.PutDimensionNodeID(ctx, instance.DbModel().InstanceID, dimension); err != nil {
+				err = errors.Wrap(err, "DatasetAPICli.PutDimensionNodeID returned an error")
+				log.Event(ctx, err.Error(), log.Error(err), log.Data{"instance_id": instance.DbModel().InstanceID, "dimension_id": dbDimension.DimensionID})
+				problem <- err
+				return
+			}
 
-		if err = hdlr.DatasetAPICli.PutDimensionNodeID(ctx, instance.DbModel().InstanceID, dimension); err != nil {
-			return errors.Wrap(err, "DatasetAPICli.PutDimensionNodeID returned an error")
-		}
+			// todo: remove this temp hack once the time codelist / input data has been fixed.
+			if dbDimension.DimensionID != "time" {
+				if err = hdlr.Store.CreateCodeRelationship(context.Background(), instance.DbModel().InstanceID, dimension.CodeListID(), dbDimension.Option); err != nil {
+					err = errors.Wrap(err, "error attempting to create relationship to code")
+					log.Event(ctx, err.Error(), log.Error(err), log.Data{"instance_id": instance.DbModel().InstanceID, "dimension_id": dbDimension.DimensionID})
+					problem <- err
+					return
+				}
+			}
+		}(dimension)
 
-		// todo: remove this temp hack once the time codelist / input data has been fixed.
-		if dbDimension.DimensionID != "time" {
-			if err = hdlr.Store.CreateCodeRelationship(ctx, instance.DbModel().InstanceID, dimension.CodeListID(), dbDimension.Option); err != nil {
-				return errors.Wrap(err, "error attempting to create relationship to code")
+		if count >= hdlr.BatchSize {
+			wg.Wait()
+			if len(problem) > 0 {
+				return <-problem
 			}
+			count = 0
```

Comment:
This code means you have to wait for a batch to finish before you can check if its been successful. I think it might be possible to bail out sooner with something like:

```
var batchProcessed chan bool

if count >= hdlr.BatchSize {
	processBatch(batchProcessed)

	select {
	case batchProcessed:
		count = 0
	case err := <- problem:
		return err
	}
}

...

func processBatch(batchProcessed chan bool) {
	go func() {
		wg.Wait()
		batchProcessed <- true
	}()
}
```
Here you are waiting on either the wait group to complete or the problem channel to return en error. **I haven't tested this and I could be completely wrong but please feel free to correct me.**

Reply:
Totally valid point, and a good spot. I have changed the code as suggested. I extracted the function that waits for the batch to be complete / failed, so that I can also apply the same logic on the final batch.